cloudevents/working-drafts/xml-format.md

```diff
@@ -0,0 +1,367 @@
+
+* The [Batch](#5-xml-batch-format) section describes how multiple CloudEvents
+can be packaged into a single XML document.
+
+### 1.1. Conformance
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC2119][rfc2119].
+
+### 1.2 Approach
+
+The XML representation used is deliberately verbose in favor of readability
+and deterministic processing.
+
+Preservation of type information for CloudEvent attributes is supported allowing
+custom extensions to be communicated without type loss.
+
+A schema-less approach has been taken favoring convention over rigid document
+structure.
+
+The namespace `http://cloudevents.io/xmlformat/V1` MUST be used.
+
+When namespace prefixes are used a prefix of `ce` is preferred but MUST NOT
+be expected from an XML document processing perspective.
+
+An XML document preamble SHOULD be included to ensure deterministic processing.
+
+## 2. Attributes
+
+The CloudEvents type system is mapped to a set of CloudEvent specific
+type designators as follows :
+
+| CloudEvents Type  |  XML Format Type | XML Schema Type | Notes |
+| :-----------------| :----------------| :--------------- | :---- |
+| Boolean       | ce:boolean | [xs:boolean][xml-primitives] | `true` or `false` |
+| Integer       | ce:int | [xs:int][xml-primitives] | |
+| String        | ce:string | [xs:string][xml-primitives] | |
+| Binary        | ce:binary | [xs:base64Binary][xml-primitives] |  |
+| URI           | ce:uri | [xs:anyURI][xml-primitives] following [RFC 3986][rfc3986]| |
+| URI-reference | ce:uriRef | [xs:anyURI][xml-primitives] following [RFC 3986][rfc3986] | |
+| Timestamp     | ce:timestamp | [xs:dateTime][xml-primitives] following [RFC 3339][rfc3339] | |
+
+Each CloudEvent context attribute is represented as an XML element whose local
+name exactly matches that of the attribute.
+
+See the [envelope](#4-envelope) for special handing of the `specversion` context attribute.
+
+Extension attributes MUST be decorated with the appropriate CloudEvent format
+designators using using the `xsi:type` XML attribute, this allows them to be exchanged
+without loss of type information.
+
+REQUIRED and OPTIONAL context attributes SHOULD NOT be decorated with an `xsi:type`.
+
+No other XML element attributes are expected, if present they MUST be ignored during
+processing.
+
+``` xml
+    ...
+    <id>AAABBBCCCNNN0000</id>
+    ..
+    <time>2021-08-14T14:30:22-08:00</time>
+    ..
+    <myextension xsi:type="ce:string">my extension value</myextension>
+    <myboolean xsi:type="ce:boolean">false</myboolean>
+```
+
+## 3. Data
+
+The data portion of a CloudEvent follows a similar model to that employed by
+the [JSON Format specification][json-format]. A `data` element MUST be used to
+encapsulate the payload.
+
+An `xsi:type` is used to discrimate the payload type and MUST be present.
```

Comment:
I don't have a strong feeling about this but I thought I'd ask... in JSON we allow for a receiver to infer that a non-type-specified Data is JSON. Should we allow for a non-type-specified data default to XML too?

If we're going to require xsi:type on "data", which is a spec defined attribute, I'm struggling with the inconsistency above where we ban xsi:type on the other spec defined ones. I understand that "data" is special, but the inconsistency is hard to justify. Couldn't we just say that using an xsi:type that is inconsistent with the spec's definition of those attributes is an error?

Reply:
If the user wishes to infer that it's fine, but i don't see how that impacts the format specification given the purpose is to support the reliable exchange of a CE.

'data' in this situation is-not a CloudEvent attribute, it's a format specific  construct.

In the early dsays of this spec i had separate XML element names for the different data-content types, that was changed to this more XML idiomatic way of specifying type information.

Reply:
I think what @duglin was potentially suggesting was *something* like:

> If a CloudEvent doesn't specify a `datacontenttype` attribute, but has a `ce:data` element with an `xsi:type` of `xs:any`, the event should be processed as if it had a `datacontenttype` of `text/xml`

Or it's possible that I misunderstood entirely :)

Reply:
I think what @duglin was potentially suggesting was *something* like:

> If a CloudEvent doesn't specify a `datacontenttype` attribute, but has a `ce:data` element with an `xsi:type` of `xs:any`, the event should be processed as if it had a `datacontenttype` of `text/xml`

Or it's possible that I misunderstood entirely :)

Reply:
yes that was part of what I was trying to say.

the other part is that I'm not sure why we're banning xsi:type on spec defined attributes. Seems like an unnecessary restriction, especially for folks who like to be explicit about things

Reply:
> the other part is that I'm not sure why we're banning xsi:type on spec defined attributes. Seems like an unnecessary restriction, especially for folks who like to be explicit about things

Agreed - my implementation currently allows this, but requires that the xsi:type matches the one in the spec. (You can't have an event with an ID that happens to be "true" and declare it with `<ce:id xsi:type="ce:boolean">true</ce:id>` for example.)

Reply:
Agreed

Reply:
OK ... 

Will change the context-attribute commentary to reflect the fact that xsi:type MAY be present for non-extension attributes, if present the xi:type MUST match that of the expected type for the CE attribute.

I'm not a fan of coercing the data on reception, imho datacontenttype is only used during serialization, or deserialization you either get (1) Binary, (2) Text, or (3) XML; anything other than that is for a higher-order function to decide.

Reply:
I believe this is addressed now.